src/box/space.c

```diff
@@ -431,6 +431,70 @@ space_delete(struct space *space)
 	space->vtab->destroy(space);
 }
 
+int
+space_foreach(int (*func)(struct space *sp, void *udata), void *udata)
+{
+	struct space *space = space_by_id(BOX_SPACE_ID);
+	if (space == NULL)
+		return -1;
+	struct index *pk = space_index(space, 0);
+	if (pk == NULL)
+		return -1;
+	int rc = -1;
+	char key[6];
+	struct tuple *tuple;
+	/*
+	 * Step 1: iterate over system spaces.
+	 */
+	assert(mp_sizeof_uint(BOX_SYSTEM_ID_MIN) <= sizeof(key));
+	mp_encode_uint(key, BOX_SYSTEM_ID_MIN);
+	struct iterator *it = index_create_iterator(pk, ITER_GE, key, 1);
+	if (it == NULL)
+		return -1;
+	while ((rc = iterator_next(it, &tuple)) == 0 && tuple != NULL) {
+		uint32_t id;
+		if (tuple_field_u32(tuple, BOX_SPACE_FIELD_ID, &id) != 0)
+			continue;
+		if (id > BOX_SYSTEM_ID_MAX)
+			break;
+		space = space_cache_find(id);
+		if (space == NULL)
+			break;
+		rc = func(space, udata);
+		if (rc != 0)
+			break;
+	}
+	iterator_delete(it);
+	if (rc != 0)
+		return -1;
+	/*
+	 * Step 2: iterate over non-system spaces.
+	 * The second traverse is required, because spaces with ids in the range
+	 * [0..BOX_SYSTEM_ID_MIN] are actually non-system.
+	 */
+	mp_encode_uint(key, 0);
+	it = index_create_iterator(pk, ITER_GE, key, 1);
+	if (it == NULL)
+		return -1;
+	while ((rc = iterator_next(it, &tuple)) == 0 && tuple != NULL) {
```

Comment:
Nit: It'd probably look better (no code duplication) if we had a helper function that would iterate over a space range:

```C
int
space_foreach(...)
{
    space_foreach_helper(0, BOX_SYSTEM_ID_MIN, cb, arg);
    space_foreach_helper(BOX_SYSTEM_ID_MAX + 1, BOX_SPACE_MAX, cb, arg);
}
```

Reply:
Changed to `space_foreach_helper` (3 times).